File: 3.x.x/crates-utopia/cli/src/adapter.rs
/// Runtime that executes the CLI dispatch callback.
///
/// PHP `Utopia\CLI\Adapter`.
pub trait Adapter: Send + std::fmt::Debug {
    /// PHP public `$workerNum`.
    fn worker_num(&self) -> i32;

    /// PHP `$adapter->start($callback)`.
    fn start(&mut self, callback: &mut dyn FnMut());

    /// PHP `$adapter->stop()`.
    fn stop(&mut self);

    /// PHP `$adapter->onWorkerStart($callback)`.
    fn on_worker_start(&mut self, callback: WorkerCallback);

    /// PHP `$adapter->onWorkerStop($callback)`.
    fn on_worker_stop(&mut self, callback: WorkerCallback);

    /// PHP `$adapter->onJob($callback)`.
    fn on_job(&mut self, callback: &mut dyn FnMut());
}

/// Worker lifecycle callback. PHP receives `($pool, $workerId)`; Rust gets the worker id.
pub type WorkerCallback = Box<dyn Fn(String) + Send + Sync>;

File: packages/cli/src/CLI/Adapter.php
<?php

declare(strict_types=1);

namespace Utopia\CLI;

abstract class Adapter
{
    public int $workerNum;

    public function __construct(int $workerNum = 0)
    {
        $this->workerNum = $workerNum;
    }

    /**
     * Starts the Server.
     *
     * @param $callback
     */
    abstract public function start($callback): self;

    /**
     * Stops the Server.
     */
    abstract public function stop(): self;

    /**
     * Is called when a Worker starts.
     */
    abstract public function onWorkerStart(callable $callback): self;

    /**
     * Is called when a Worker stops.
     */
    abstract public function onWorkerStop(callable $callback): self;

    /**
     * Is called when a job is processed.
     */
    abstract public function onJob(callable $callback): self;
}
